src/compiler/checker.ts

```diff
@@ -20275,28 +20275,35 @@ namespace ts {
                         }
                         const sourceTypeArguments = getTypeArguments(source);
                         const targetTypeArguments = getTypeArguments(target);
-                        const startCount = Math.min(isTupleType(source) ? getStartElementCount(source.target, ElementFlags.NonRest) : 0, getStartElementCount(target.target, ElementFlags.NonRest));
-                        const endCount = Math.min(isTupleType(source) ? getEndElementCount(source.target, ElementFlags.NonRest) : 0, targetRestFlag ? getEndElementCount(target.target, ElementFlags.NonRest) : 0);
+                        const targetStartCount = getStartElementCount(target.target, ElementFlags.NonRest);
+                        const targetEndCount = getEndElementCount(target.target, ElementFlags.NonRest);
+                        const targetHasRestElement = target.target.hasRestElement;
                         let canExcludeDiscriminants = !!excludedProperties;
-                        for (let i = 0; i < targetArity; i++) {
-                            const sourceIndex = i < targetArity - endCount ? i : i + sourceArity - targetArity;
-                            const sourceFlags = isTupleType(source) && (i < startCount || i >= targetArity - endCount) ? source.target.elementFlags[sourceIndex] : ElementFlags.Rest;
-                            const targetFlags = target.target.elementFlags[i];
+                        for (let sourcePosition = 0; sourcePosition < sourceArity; sourcePosition++) {

```

Comment:
The core of the fix is in the iteration "style". Previously we were iterating over `targetArity` and "slicing" a source in the position of the target rest element. 

When we consider the added test case:
```ts
declare let tt3: [number, string, ...any[]]
let tt4: [number, ...number[]] = tt3
```
It means that for the target's `...number[]` we were creating a slice from those elements: `string, ...any[]`. This, in turn, was indexed~/unionized and thus became `string | any` which is just an equivalent of `any`. And that was used as a source type for the target, introducing a bug.

I've flipped this and now this loop iterates over `sourceArity`. This ensures that a specific position in the source is assignable, in "isolation", to the respective target position.

Reply:
From testing this code with some examples and carefully reading through it, I think it's at least as correct as before (and more, now that it fixes the bug). I still need to review the error messages though.

Reply:
Thank you very much for the review! I know that there are some changes in the error messages and stuff - just want to let you know that I would happily make any required adjustments to this PR, but when it comes to those error messages I need to know what would be the expected results. IIRC the current messages are not incorrect - but perhaps you'd like to have them improved somehow.